File: lib/entities/device_request.dart
enum DeviceRequestStatus {
  pending,
  approved,
  rejected;

  static DeviceRequestStatus fromString(String? value) {
    switch (value?.toUpperCase()) {
      case 'APPROVED':
        return DeviceRequestStatus.approved;
      case 'REJECTED':
        return DeviceRequestStatus.rejected;
      default:
        return DeviceRequestStatus.pending;
    }
  }

  String get value {
    switch (this) {
      case DeviceRequestStatus.pending:
        return 'PENDING';
      case DeviceRequestStatus.approved:
        return 'APPROVED';
      case DeviceRequestStatus.rejected:
        return 'REJECTED';
    }
  }

  String get label {
    switch (this) {
      case DeviceRequestStatus.pending:
        return 'Đang chờ';
      case DeviceRequestStatus.approved:
        return 'Đã duyệt';
      case DeviceRequestStatus.rejected:
        return 'Từ chối';
    }
  }
}

class DeviceRequest {
  final String id;
  final String deviceCode;
  final String? deviceName;
  final String? roomId;
  final String? roomName;
  final String? requestedBy;
  final String? requesterName;
  final DeviceRequestStatus status;
  final String? reviewedBy;
  final String? reviewerName;
  final DateTime? reviewedAt;
  final String? adminNote;
  final DateTime createdAt;

  const DeviceRequest({
    required this.id,
    required this.deviceCode,
    this.deviceName,
    this.roomId,
    this.roomName,
    this.requestedBy,
    this.requesterName,
    this.status = DeviceRequestStatus.pending,
    this.reviewedBy,
    this.reviewerName,
    this.reviewedAt,
    this.adminNote,
    required this.createdAt,
  });

  factory DeviceRequest.fromJson(Map<String, dynamic> json) {
    return DeviceRequest(
      id: json['id'] as String,
      deviceCode: json['device_code'] as String,
      deviceName: json['device_name'] as String?,
      roomId: json['room_id'] as String?,
      roomName: json['room_name'] as String?,
      requestedBy: json['requested_by'] as String?,
      requesterName: json['requester_name'] as String?,
      status: DeviceRequestStatus.fromString(json['status'] as String?),
      reviewedBy: json['reviewed_by'] as String?,
      reviewerName: json['reviewer_name'] as String?,
      reviewedAt: json['reviewed_at'] != null
          ? DateTime.parse(json['reviewed_at'] as String)
          : null,
      adminNote: json['admin_note'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'device_code': deviceCode,
      'device_name': deviceName,
      'room_id': roomId,
    };
  }

  DeviceRequest copyWith({
    String? id,
    String? deviceCode,
    String? deviceName,
    String? roomId,
    String? roomName,
    String? requestedBy,
    String? requesterName,
    DeviceRequestStatus? status,
    String? reviewedBy,
    String? reviewerName,
    DateTime? reviewedAt,
    String? adminNote,
    DateTime? createdAt,
  }) {
    return DeviceRequest(
      id: id ?? this.id,
      deviceCode: deviceCode ?? this.deviceCode,
      deviceName: deviceName ?? this.deviceName,
      roomId: roomId ?? this.roomId,
      roomName: roomName ?? this.roomName,
      requestedBy: requestedBy ?? this.requestedBy,
      requesterName: requesterName ?? this.requesterName,
      status: status ?? this.status,
      reviewedBy: reviewedBy ?? this.reviewedBy,
      reviewerName: reviewerName ?? this.reviewerName,
      reviewedAt: reviewedAt ?? this.reviewedAt,
      adminNote: adminNote ?? this.adminNote,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  bool get isPending => status == DeviceRequestStatus.pending;
  bool get isApproved => status == DeviceRequestStatus.approved;
  bool get isRejected => status == DeviceRequestStatus.rejected;

  @override
  String toString() =>
      'DeviceRequest(id: $id, code: $deviceCode, status: ${status.value})';
}

File: backend/app/schemas/v1/device_request.py
from __future__ import annotations
import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


class DeviceRequestCreate(BaseModel):
    """POST /api/v1/device-requests — submit a device permission request."""
    device_code: str = Field(..., max_length=50)
    device_name: Optional[str] = Field(None, max_length=100)
    room_id: Optional[uuid.UUID] = None  # null = global


class DeviceRequestSubmit(BaseModel):
    """Submit a new device permission request (tablet/admin app)."""
    device_code: str = Field(..., max_length=50)
    device_name: Optional[str] = Field(None, max_length=100)
    room_id: Optional[uuid.UUID] = None


class ApproveRequest(BaseModel):
    """PATCH /api/v1/device-requests/{id}/approve — approve a device request."""
    admin_note: Optional[str] = Field(None, max_length=255)


class RejectRequest(BaseModel):
    """PATCH /api/v1/device-requests/{id}/reject — reject a device request."""
    admin_note: Optional[str] = Field(None, max_length=255)


class DeviceRequestResponse(BaseModel):
    """GET /api/v1/device-requests/{id} — single device request."""
    id: uuid.UUID
    device_code: str
    device_name: Optional[str]
    room_id: Optional[uuid.UUID]
    room_name: Optional[str] = None
    requested_by: Optional[uuid.UUID]
    requester_name: Optional[str] = None
    status: Literal["PENDING", "APPROVED", "REJECTED"]
    reviewed_by: Optional[uuid.UUID]
    reviewer_name: Optional[str] = None
    reviewed_at: Optional[datetime]
    admin_note: Optional[str]
    created_at: datetime

    model_config = {"from_attributes": True}


class DeviceRequestList(BaseModel):
    """GET /api/v1/device-requests — paginated list."""
    total: int
    items: list[DeviceRequestResponse]
